File: ios/Reguerta/Reguerta/Domain/Orders/ReceivedOrdersModels.swift
import Foundation

enum ProducerOrderStatus: String, CaseIterable, Equatable, Sendable {
    case unread
    case read
    case prepared
    case delivered

    static func from(_ rawValue: String?) -> ProducerOrderStatus {
        guard let rawValue else { return .unread }
        return ProducerOrderStatus(rawValue: rawValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()) ?? .unread
    }

    var title: String {
        switch self {
        case .unread: return "Pendiente"
        case .read: return "Pendiente"
        case .prepared: return "Preparado"
        case .delivered: return "Entregado"
        }
    }
}

struct ReceivedOrderLineRecord: Identifiable, Equatable, Sendable {
    let id: String
    let orderId: String
    let consumerId: String
    let consumerDisplayName: String
    let productId: String
    let productName: String
    let productImageUrl: String?
    let companyName: String
    let packagingLine: String
    let quantity: Double
    let quantityUnitSingular: String
    let quantityUnitPlural: String
    let measureQuantityPerUnit: Double
    let measureUnitSingular: String
    let measureUnitPlural: String
    let measureUnitAbbreviation: String?
    let isWeightPricing: Bool
    let subtotal: Double

    var dedupKey: String {
        "\(orderId)|\(consumerId)|\(productId)"
    }

    var totalMeasureQuantity: Double {
        guard isWeightPricing else {
            return quantity * measureQuantityPerUnit
        }
        return weightQuantityRepresentsMeasure ? quantity : quantity * measureQuantityPerUnit
    }

    var orderedQuantity: Double {
        guard isWeightPricing, measureQuantityPerUnit > 0 else {
            return quantity
        }
        return weightQuantityRepresentsMeasure ? quantity / measureQuantityPerUnit : quantity
    }

    private var weightQuantityRepresentsMeasure: Bool {
        guard isWeightPricing, measureQuantityPerUnit > 0 else {
            return true
        }
        return quantity >= measureQuantityPerUnit
    }
}

struct ReceivedOrdersProductRow: Identifiable, Equatable, Sendable {
    let productId: String
    let productName: String
    let productImageUrl: String?
    let companyName: String
    let packagingLine: String
    let totalQuantity: Double
    let quantityUnitSingular: String
    let quantityUnitPlural: String
    let totalMeasureQuantity: Double
    let measureUnitSingular: String
    let measureUnitPlural: String
    let measureUnitAbbreviation: String?
    let subtotal: Double

    var id: String { productId }

    func quantityUnitLabel() -> String {
        receivedOrdersIsApproximatelyOne(totalQuantity) ? quantityUnitSingular : quantityUnitPlural
    }

    func totalMeasureLabel() -> String {
        receivedOrdersMeasureLabel(
            quantity: totalMeasureQuantity,
            singular: measureUnitSingular,
            plural: measureUnitPlural,
            abbreviation: measureUnitAbbreviation,
            prefersAbbreviation: true
        )
    }
}

struct ReceivedOrdersMemberLine: Identifiable, Equatable, Sendable {
    let id: String
    let productName: String
    let packagingLine: String
    let quantity: Double
    let quantityUnitSingular: String
    let quantityUnitPlural: String
    let totalMeasureQuantity: Double
    let measureUnitSingular: String
    let measureUnitPlural: String
    let measureUnitAbbreviation: String?
    let subtotal: Double

    func quantityUnitLabel() -> String {
        receivedOrdersIsApproximatelyOne(quantity) ? quantityUnitSingular : quantityUnitPlural
    }

    func totalMeasureLabel() -> String {
        receivedOrdersMeasureLabel(
            quantity: totalMeasureQuantity,
            singular: measureUnitSingular,
            plural: measureUnitPlural,
            abbreviation: measureUnitAbbreviation,
            prefersAbbreviation: true
        )
    }
}

struct ReceivedOrdersMemberGroup: Identifiable, Equatable, Sendable {
    let id: String
    let orderId: String
    let consumerDisplayName: String
    let producerStatus: ProducerOrderStatus
    let lines: [ReceivedOrdersMemberLine]
    let total: Double
}

struct ReceivedOrdersSnapshot: Equatable, Sendable {
    let byProductRows: [ReceivedOrdersProductRow]
    let byMemberGroups: [ReceivedOrdersMemberGroup]
    let generalTotal: Double
}

func receivedOrdersIsApproximatelyOne(_ value: Double) -> Bool {
    abs(value - 1) < 0.000_1
}

func receivedOrdersMeasureLabel(
    quantity: Double,
    singular: String,
    plural: String,
    abbreviation: String?,
    prefersAbbreviation: Bool
) -> String {
    let quantityText = quantity.myOrderUiDecimal
    let numberAwareUnit = receivedOrdersIsApproximatelyOne(quantity) ? singular : plural
    let unit = prefersAbbreviation
        ? (abbreviation?.isNotEmpty == true ? abbreviation! : numberAwareUnit)
        : numberAwareUnit
    return [quantityText, unit]
        .filter(\.isNotEmpty)
        .joined(separator: " ")
}

extension ReceivedOrdersSnapshot {
    func withProducerStatus(orderId: String, status: ProducerOrderStatus) -> ReceivedOrdersSnapshot {
        ReceivedOrdersSnapshot(
            byProductRows: byProductRows,
            byMemberGroups: byMemberGroups.map { group in
                if group.orderId == orderId {
                    return ReceivedOrdersMemberGroup(
                        id: group.id,
                        orderId: group.orderId,
                        consumerDisplayName: group.consumerDisplayName,
                        producerStatus: status,
                        lines: group.lines,
                        total: group.total
                    )
                }
                return group
            },
            generalTotal: generalTotal
        )
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/orders/ReceivedOrdersModels.kt
package com.reguerta.user.domain.orders

import java.util.Locale
import kotlin.math.abs

enum class ReceivedOrderProducerStatus(val wireValue: String) {
    UNREAD("unread"),
    READ("read"),
    PREPARED("prepared"),
    DELIVERED("delivered"),
    ;

    companion object {
        fun fromWireValue(rawValue: String?): ReceivedOrderProducerStatus {
            val normalized = rawValue?.trim()?.lowercase(Locale.ROOT)
            return entries.firstOrNull { status -> status.wireValue == normalized } ?: UNREAD
        }
    }
}

data class ReceivedOrderLineRecord(
    val id: String,
    val orderId: String,
    val consumerId: String,
    val consumerDisplayName: String,
    val productId: String,
    val productName: String,
    val productImageUrl: String?,
    val companyName: String,
    val packagingLine: String,
    val quantity: Double,
    val quantityUnitSingular: String,
    val quantityUnitPlural: String,
    val measureQuantityPerUnit: Double,
    val isWeightPricing: Boolean,
    val subtotal: Double,
) {
    val dedupKey: String
        get() = "$orderId|$consumerId|$productId"

    val orderedQuantity: Double
        get() = if (isWeightPricing && measureQuantityPerUnit > 0.0 && weightQuantityRepresentsMeasure) {
            quantity / measureQuantityPerUnit
        } else {
            quantity
        }

    private val weightQuantityRepresentsMeasure: Boolean
        get() {
            if (!isWeightPricing || measureQuantityPerUnit <= 0.0) return true
            return quantity >= measureQuantityPerUnit
        }
}

data class ReceivedOrdersProductRow(
    val productId: String,
    val productName: String,
    val productImageUrl: String?,
    val packagingLine: String,
    val totalQuantity: Double,
    val quantityUnitSingular: String,
    val quantityUnitPlural: String,
) {
    fun quantityUnitLabel(): String =
        if (receivedOrdersIsApproximatelyOne(totalQuantity)) quantityUnitSingular else quantityUnitPlural
}

data class ReceivedOrdersMemberLine(
    val id: String,
    val productName: String,
    val packagingLine: String,
    val quantity: Double,
    val quantityUnitSingular: String,
    val quantityUnitPlural: String,
    val subtotal: Double,
) {
    fun quantityUnitLabel(): String =
        if (receivedOrdersIsApproximatelyOne(quantity)) quantityUnitSingular else quantityUnitPlural
}

data class ReceivedOrdersMemberGroup(
    val id: String,
    val orderId: String,
    val consumerDisplayName: String,
    val producerStatus: ReceivedOrderProducerStatus,
    val lines: List<ReceivedOrdersMemberLine>,
    val total: Double,
)

data class ReceivedOrdersSnapshot(
    val byProductRows: List<ReceivedOrdersProductRow>,
    val byMemberGroups: List<ReceivedOrdersMemberGroup>,
    val generalTotal: Double,
) {
    fun withProducerStatus(
        orderId: String,
        status: ReceivedOrderProducerStatus,
    ): ReceivedOrdersSnapshot =
        copy(
            byMemberGroups = byMemberGroups.map { group ->
                if (group.orderId == orderId) {
                    group.copy(producerStatus = status)
                } else {
                    group
                }
            },
        )
}

enum class ReceivedOrderStatusWriteResult {
    SUCCESS,
    PERMISSION_DENIED,
    FAILURE,
}

fun Throwable.toReceivedOrderStatusWriteResult(): ReceivedOrderStatusWriteResult {
    val codeName = runCatching {
        javaClass.methods.firstOrNull { method -> method.name == "getCode" }
            ?.invoke(this)
            ?.toString()
            ?.uppercase()
    }.getOrNull()
    val normalizedMessage = message?.uppercase().orEmpty()

    return if (
        codeName?.contains("PERMISSION_DENIED") == true ||
        normalizedMessage.contains("PERMISSION_DENIED") ||
        normalizedMessage.contains("PERMISSION-DENIED")
    ) {
        ReceivedOrderStatusWriteResult.PERMISSION_DENIED
    } else {
        ReceivedOrderStatusWriteResult.FAILURE
    }
}

private fun receivedOrdersIsApproximatelyOne(value: Double): Boolean =
    abs(value - 1.0) < 0.0001
